benches/nexmark/nexmark_dbsp_source.rs

```diff
@@ -87,19 +87,18 @@ where
         // Otherwise we want to emit at least one event, so if the next event
         // is still in the future, we sleep until we can emit it.
         let next_event = next_event.unwrap();
-        let wallclock_time_now = wallclock_time();
+        let mut wallclock_time_now = self.generator.wallclock_time();
         if next_event.wallclock_timestamp > wallclock_time_now {
-            sleep(Duration::from_millis(
-                next_event.wallclock_timestamp - wallclock_time_now,
-            ));
+            let millis_to_sleep = next_event.wallclock_timestamp - wallclock_time_now;
+            sleep(Duration::from_millis(millis_to_sleep));
```

Comment:
@ryzhyk this is the kind of situation where I think that being able to pause and resume operators would be really valuable, we could do computation while things like network delays, disk delays, etc. are all occurring within other operators

Reply:
I've been thinking about this lately.  Clearly, operators shouldn't block.  We discussed this with @absoludity , and the above implementation is just a temporary solution until we have proper infrastructure in place. Sleeping is kinda ok in this case because the circuit has exactly one source operator, and if the source does not have any data, the circuit can just pause (except that it could use the idle time to perform trace compaction).

We already have a way to pause the operator via the `is_async`/`ready`/`register_ready_callback` API in `trace Operator`; however pausing is not the right solution for either operators that perform I/O or operators like the one above that just need to sleep. A paused operator will prevent the entire circuit from completing the clock cycle and making forward progress.  Imagine a circuit with two source operators, one reading from file and one from a network socket.  At any given time only one of them may have data, so we shouldn't stall the circuit just because the other operator is empty.   Instead, whenever at least one of the sources has data available, the computation should go forward, with the other source returning an empty batch.  However, when both sources are empty, there is no point in keeping the circuit busy while doing no useful work, so in this case we want to pause the execution, perhaps only doing trace compaction, until at least one of the sources has new inputs available.

In summary, there are two levels of control we need exercise: within a single clock cycle the async operator mechanism can be used to tell the scheduler to postpone the evalution of an operator. As far as I can tell, this is really only useful for worker to worker communication (which is what this mechanism was created for in the first place). A different mechanism is needed to schedule the beginning of the next clock cycle, based on the availability of data across all sources or some other user-defined policy (the client application may be in the best position to decide when is the best time to start cranking the handle).

I have an idea for a solution and will write an RFC soon-ish.